tree: Replace RestrictiveReadonlyRecord with RestrictiveStringRecord (#22479)
microsoft/FluidFramework: .changeset/purple-hats-roll.md

---
"fluid-framework": minor
"@fluidframework/tree": minor
---
---

> "section": "tree"

Reviewer: Would this make more sense to go in the deprecations section?

Author: We have been putting all the tree things, including features in tree, so I think that takes precedence.

---
RestrictiveReadonlyRecord is deprecated

> `RestrictiveReadonlyRecord` was an attempt to implement a version of the built in `Record<TKey, TValue>` type that would prohibit (instead of leave unrestricted like Record does) values under keys that are to not extend `TKey`.

Reviewer 2: "that are to not extend" wording is a bit hard to understand

Author: ```suggestion
`RestrictiveReadonlyRecord` was an attempt to implement a version of the built in `Record<TKey, TValue>` type that would prohibit (instead of leaving unrestricted like Record does) values under keys that are do not extend `TKey`.
```

> `RestrictiveReadonlyRecord` was an attempt to implement a version of the built in `Record<TKey, TValue>` type that would prohibit (instead of leaving unrestricted like Record does) values under keys that are do not extend `TKey`.

Reviewer: ```suggestion
`RestrictiveReadonlyRecord` was an attempt to implement a version of the built-in `Record<TKey, TValue>` type that would prohibit (instead of leaving unrestricted like Record does) values under keys that do not extend `TKey`.
```
```suggestion
`RestrictiveReadonlyRecord` was an attempt to implement a version of the built in `Record<TKey, TValue>` type that would prohibit (instead of leaving unrestricted like Record does) values under keys that are do not extend `TKey`.
```

The implementation of `RestrictiveReadonlyRecord` failed to accomplish this except for the edge cases where `TKey` was exactly `string` or exactly `symbol`.

> Fixing this bug appears to be impossible withing the current limitation of TypeScript, however this library does not require any case other than `TKey` being exactly `string`.

Reviewer 3: ```suggestion
Fixing this bug appears to be impossible within the current limitation of TypeScript, however this library does not require any case other than `TKey` being exactly `string`.
```

> To reduce the risk of users of this the tree library using the problematic `RestrictiveReadonlyRecord` type it has been deprecated and replaced with a more specific type that avoids the bug, `RestrictiveStringRecord<TValue>`.

Reviewer: ```suggestion
To reduce the risk of users of the tree library using the problematic `RestrictiveReadonlyRecord` type, it has been deprecated and replaced with a more specific type that avoids the bug, `RestrictiveStringRecord<TValue>`.
```
```suggestion
To reduce the risk of users of this the tree library using the problematic `RestrictiveReadonlyRecord` type it has been deprecated and replaced with a more specific type that avoids the bug, `RestrictiveStringRecord<TValue>`.
```

> To highlight that this new type is not intended for direct use by users of tree, and instead is just used as part of the typing of its public API, `RestrictiveStringRecord` as been tagged with `@system`.

Reviewer 2: ```suggestion
To highlight that this new type is not intended for direct use by users of tree, and instead is just used as part of the typing of its public API, `RestrictiveStringRecord` has been tagged with `@system`.
```
